javascript/react-js/react-testing-part-two.md

```diff
@@ -0,0 +1,155 @@
+### Introduction
+
+We've already covered some basics of React testing. Now's the time to dive deeper. 
+
+In this lesson, we'll learn about mocking. Furthermore, we'll discuss a React component from this ([theodinproject.com](https://theodinproject.com)) website and understand how React tests are written in a real world application. To top it all off, we'll tackle the practice exercise at the end of this lesson to solidify what we've learned. 
+
+### Learning Outcomes
+
+By the end of this lesson, you should be able to do the following:
+
+TODO
+
+### What is Mocking?
+
+If you've been following along our lessons, the concept of mocking has already been introduced in the earlier section and you might have even incorporated mocks in your [Battleship](https://www.theodinproject.com/paths/full-stack-javascript/courses/javascript/lessons/battleship) project. Let's look at how mocks'll help in testing React components.
+
+#### Testing Callback Handlers
+
+Callbacks are ubiquitous. Every avenue of user interaction involves callbacks. Sometimes they're passed in as props to alter state of the parent component. Consider this simple input component:
+
+~~~javascript
+// FavoriteInput.js
+
+import React from "react";
+
+const FavoriteInput = ({ onChange, id }) => (
+  <label htmlFor={id}>
+    What is your favorite wild animal?
+    <input id={id} onChange={onChange} />
+  </label>
+);
+
+export default FavoriteInput;
+~~~
+
+Nothing fancy. `FavoriteInput` is a simple component with a couple props passed in. We're interested in the `onChange` prop. We have no idea what the function does. We have no idea how the function will affect the application. All we know is it must be called when user types in the input box. Let's test it. 
+
+~~~javascript
+// FavoriteInput.test.js
+
+import React from "react";
+import { render } from "@testing-library/react";
+import "@testing-library/jest-dom";
+import userEvent from "@testing-library/user-event";
+import FavoriteInput from "./FavoriteInput";
+
+describe("Favorite Input", () => {
+  let onChange;
+  let getByRole;
+  let input;
+
+  beforeEach(() => {
+    onChange = jest.fn();
+    ({ getByRole } = render(<FavoriteInput onChange={onChange} />));
+    input = getByRole("textbox");
+  });
+
+  test("calls onChange correct number of times", () => {
+    userEvent.type(input, "Lion");
+    expect(onChange).toHaveBeenCalledTimes(4);
+  });
+
+  test("calls onChange with correct argument", () => {
+    userEvent.type(input, "Ox");
+
+    expect(onChange.mock.calls[0][0].target.value).toMatch("O");
+    expect(onChange.mock.calls[1][0].target.value).toMatch("x");
```

Comment:
I'm afraid I don't quite understand how to use `toHaveBeenCalledWith` or `toHaveBeenNthCalledWith` here because the `onChange` mock function gets called with the event object. How will we access the `target.value` property from there?

Reply:
:thinking: you bring up a good counterpoint. I feel like you could do something with `expect.objectContaining` but that probably wouldn't make for a good example either. Let me do some playing around with this and cirlce back to you.